app/coffee/Features/Compile/ClsiCookieManager.coffee

```diff
@@ -0,0 +1,62 @@
+Settings = require "settings-sharelatex"
+request = require('request')
+redis = require("redis-sharelatex")
+rclient = redis.createClient(Settings.redis.web)
+Cookie = require('cookie')
+logger = require "logger-sharelatex"
+
+buildKey = (project_id)->
+	return "clsiserver:#{project_id}"
+
+
+ONE_WEEK_IN_SECONDS = 60 * 60 * 24 * 7
+
+module.exports = ClsiCookieManager =
+
+	_getServerId : (project_id, callback = (err, serverId)->)->
+		multi = rclient.multi()
+		multi.get buildKey(project_id)
+		multi.exec (err, results)->
+			if err?
+				return callback(err)
+			serverId = results[0]
+			if serverId?
+				return callback(null, serverId)
+			else
+				return ClsiCookieManager._populateServerIdViaRequest project_id, callback
+
+
+	_populateServerIdViaRequest :(project_id, callback = (err, serverId)->)->
+		url = "#{Settings.apis.clsi.url}/project/#{project_id}/status"
+		request.get url, (err, res, body)->
+			if err?
+				logger.err err:err, project_id:project_id, "error getting initial server id for project"
+				return callback(err)
+			ClsiCookieManager.setServerId project_id, res, (err)->
+				if err?
+					logger.err err:err, project_id:project_id, "error setting server id via populate request"
+				callback(err)
```

Comment:
we should pass back the serverId in the callback here 

      ClsiCookieManager.setServerId project_id, res, (err,SERVERID)->
 				if err?
 					logger.err err:err, project_id:project_id, "error setting server id via populate request"
 				callback(err,SERVERID)

and pass it back from setServerId() too